### PR TITLE
Feature/dwr 1215

### PR DESCRIPTION
**Pull request by agorina**
Task list:
- [x] combine ica and summative into one fact table 
- [x] change ica_asmt to just asmt and include all asmt types 
- [x] add asmt labels 
- [x] add school_year 
- [x] convert active_asmt into a table and load it during the migrate 
- [ ] account for the latest exam per student/school year/asmt type
- [ ] handle non-scored tests: do not load them into redshift
- [ ] add school and district groups to support permissions
- [ ] migrate embargo data
- [ ] and anything else..?

### CODE REVIEW

File: common/src/main/java/org/opentestsystem/rdw/ingest/common/repository/JdbcSqlListExecutionRepository.java [hunk 1]
```diff
@@ -18,7 +18,7 @@ public JdbcSqlListExecutionRepository(final NamedParameterJdbcTemplate jdbcTempl
         this.jdbcTemplate = jdbcTemplate;
     }
 
-   @Transactional
+    @Transactional
```
Comment (agorina) [inline common/src/main/java/org/opentestsystem/rdw/ingest/common/repository/JdbcSqlListExecutionRepository.java]: unrelated.

File: migrate-olap/src/main/java/org/opentestsystem/rdw/ingest/migrate/olap/step/OlapReportingMigrateStepsConfig.java [hunk 2]
```diff
@@ -33,21 +33,22 @@
...
+    static final String loadCodesToStagingStep = "loadCodesToStagingStep";
+    static final String loadEntitiesToStagingStep = "loadEntitiesToStagingStep";
+    static final String deriveActiveAsmtStep = "deriveActiveAsmtStep";
+
+    static final String deleteCodesStepName = "deleteCodesStep";
+    static final String upsertCodesStepName = "upsertCodesStep";
+    static final String deleteEntitiesStepName = "deleteEntitiesStep";
+    static final String upsertStudentsStepName = "upsertStudentsStep";
+    static final String upsertOrganizationsStepName = "upsertOrganizationsStep";
+    static final String upsertAsmtsStepName = "upsertAsmtsStep";
```
Comment (agorina) [inline migrate-olap/src/main/java/org/opentestsystem/rdw/ingest/migrate/olap/step/OlapReportingMigrateStepsConfig.java]: Fixed IntelliJ warning.

File: migrate-olap/src/main/java/org/opentestsystem/rdw/ingest/migrate/olap/step/SqlListWithParamExecutionStep.java [hunk 3]
```diff
@@ -35,9 +32,4 @@ public RepeatStatus execute(final StepContribution contribution, final ChunkCont
         repository.execute(sqls, migrate);
         return RepeatStatus.FINISHED;
     }
-
-    @VisibleForTesting
-    Iterable<String> getSqls() {
-        return newArrayList(sqls);
-    }
```
Comment (agorina) [inline migrate-olap/src/main/java/org/opentestsystem/rdw/ingest/migrate/olap/step/SqlListWithParamExecutionStep.java]: This was not used.

File: migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml [hunk 4]
```diff
@@ -175,19 +193,20 @@ sql:
...
                   sa.subject_id,
                   sa.name,
+                  sa.label,
                   sa.type_id,
                   sa.update_import_id,
                   sa.migrate_id
                 FROM staging_asmt sa
                   LEFT JOIN asmt ra ON ra.id = sa.id
-                WHERE ra.id IS NULL and sa.deleted = 0 AND sa.type_id = 1;
+                WHERE ra.id IS NULL and sa.deleted = 0;
```
Comment (agorina) [inline migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml]: I removed this `and` because it seemed redundant. The script that extracts data makes sue that it only extracts the supported types.

File: migrate-olap/src/main/resources/application.warehouse.to.staging.sql.yml [hunk 5]
```diff
@@ -117,6 +118,25 @@ sql:
...
                 COMPUPDATE OFF
 
+        # ------------ school_year  -------------------------------------------------------------------
+        school_year:
+          sql:
+            warehouseRead: >-
+                SELECT
+                  year
+                FROM school_year
+                INTO OUTFILE S3 '${archive.root}/${migrate.aws.location}/year'
```
Comment (agorina) [inline migrate-olap/src/main/resources/application.warehouse.to.staging.sql.yml]: I do not know why, but if i name this file 'schoo_year` it conflicts with the `school` file down below.

File: migrate-olap/src/main/resources/application.warehouse.to.staging.sql.yml [hunk 6]
```diff
@@ -252,11 +272,12 @@ sql:
                school_year,
                subject_id,
                name,
+               label,
                deleted,
                update_import_id,
                ? as migrate_id
              FROM asmt wa
-              WHERE wa.type_id = 1
+              WHERE wa.type_id in (1, 3)
```
Comment (agorina) [inline migrate-olap/src/main/resources/application.warehouse.to.staging.sql.yml]: Now support ICA and Summative.

File: migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml [hunk 7]
```diff
@@ -116,6 +116,24 @@ sql:
                DELETE FROM grade
                WHERE NOT EXISTS(SELECT sg.id FROM staging_grade sg WHERE sg.id = grade.id);
 
+       # ------------ school_year  -------------------------------------------------------------------
+        school_year:
+          sql:
+            update:  >-
+               SELECT 1;
```
Comment (jtreuting) [inline migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml]: is this just a placeholder, or does it have a purpose?

File: migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml [hunk 8]
```diff
@@ -447,9 +467,24 @@ sql:
...
+      # ------------ active_asmt_per_year -----------------------------------------------------------------------
+        active_asmt_per_year:
+          sql:
+            insert: >-
+             INSERT INTO active_asmt_per_year (asmt_id, school_year)
+               SELECT asmt_id, school_year
+                 FROM
+                 ( SELECT DISTINCT f.asmt_id as asmt_id,  f.asmt_grade_id AS grade_id, f.school_year,  asmt.subject_id, asmt.type_id FROM fact_student_exam f JOIN asmt asmt ON asmt.id = f.asmt_id
+                   UNION
+                   SELECT id AS asmt_id, grade_id, school_year, subject_id, type_id FROM asmt
```
Comment (jtreuting) [inline migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml]: it seems like the part after the `UNION` would cover everything and maybe the first part isn't needed?

the first part is all assessments where a student has taken an exam for them, and the second part is all assessments in general, right?  if a student took an exam for an assessment doesn't that mean it's in the asmt table already so that covers both situations?

i think there is a distinct chance i'm missing some part of this equation though.

also small nit: seems odd to have the alias for the `asmt` table be `asmt`, couldn't that just be left off?

File: migrate-olap/src/main/java/org/opentestsystem/rdw/ingest/migrate/olap/step/OlapReportingMigrateStepsConfig.java [hunk 9]
```diff
@@ -124,17 +125,29 @@ public Step upsertAsmtsStep() {
...
 
-        return stepBuilderFactory.get(upsertFactStudentIcaExamStepName)
+        return stepBuilderFactory.get(upsertFactStudentExamStepName)
                 .tasklet(new SqlListExecutionStep(
                         reportingSqlListExecutionRepository,
                         sqlsBuilder.build())).build();
     }
 
+    @Bean
+    public Step deriveActiveAsmtStep() {
```
Comment (malaffoon) [inline migrate-olap/src/main/java/org/opentestsystem/rdw/ingest/migrate/olap/step/OlapReportingMigrateStepsConfig.java]: nit: other steps have db-related verb in them (e.g. upsert, truncate); perhaps this could be `stageAsmtActiveYearStep` or some such. Yeah, i won't lose sleep over it either way.

File: migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml (hunk 8 above)
Comment (malaffoon) [reply migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml]: The union is needed because there may be a test result for an assessment for the "wrong" year. For example, the 2016-2017 assessment (school year = 2017) is given in 2017-2018; so we want to know that that assessment was "active" for school year 2018.

File: migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml (hunk 7 above)
Comment (agorina) [reply migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml]: There is one step to manage all CODEs and it expects all three statements (insert, update, delete) per each code. This is an empty filler for the 'update'.

File: migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml (hunk 8 above)
Comment (agorina) [reply migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml]: Will fix alias.

File: migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml (hunk 7 above)
Comment (malaffoon) [reply migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml]: And `SELECT 1` works fine in the execute method?

Comment (agorina) [reply migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml]: It seems to work. It is being used in the reporting (not olap version) this way too.

File: migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml (hunk 8 above)
Comment (jtreuting) [reply migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml]: Does it make sense (from a performance perspective) to narrow the first part to just the "off-year" results so that the distinct has less work to do?

Comment (agorina) [reply migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml]: Redshift seems to be very good with handling DISTINCT. 
The above statement with 1,217,410 exams takes about 2 sec. And with 24,215,377 records about 10 sec. 

Considering that it is going to be executed once per day, I am not sure it is worth optimizing. Let me know if you disagree.

Comment (jtreuting) [reply migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml]: i guess it depend on how fast it would be eliminating those :)  it's probably not an issue because it runs once a day

Comment (agorina) [reply migrate-olap/src/main/resources/application.staging.to.reporting.sql.yml]: I am going to keep an eye on the overhead this step adds, and if it becomes significant - will look into optimizing.

**Comment by agorina** [comment]
I am reloading the dev version of Redshift and I am seeing something puzzling:
Here are the execution time in seconds for each step:
Avg Time in sec | Step name
-|-
15.0862 | extractEntitiesStep
12.6379 | loadEntitiesToStagingStep
11.8596 |	upsertCodesStep
5.5088 |	stepTruncateStageAfter
5.1897 |	stepTruncateStageBefore
5.0000 |	upsertFactStudentExamStep
3.3860 |	upsertStudentsStep
3.2982 |	deleteEntitiesStep
2.6842 |	upsertOrganizationsStep
2.2241 |	cleanupLandingZone
1.7895 |	upsertAsmtsStep
1.2069 |	loadCodesToStagingStep
1.0526 |	deriveActiveAsmtStep
1.0526 |	deleteCodesStep
1.0000 |	extractCodesStep

Here is the total sum of all steps execution time:
```sql
select  sum(timestampdiff(SECOND, START_TIME, END_TIME)) as runtime_in_sec from BATCH_STEP_EXECUTION b ;
```
-- 4460

But the overall total JOB execution time (the same value from both `migrate` and `BATCH_JOB_EXECUTION`)

-- 6184

Where does the time go? My first though was that maybe finding the next batch takes that long, but the Aurora slow query log does not prove it.

So far I have migrates 1,465,442 exams and it has been running for 105 min. I think it is two times slower than I remember from the last run (but I may have increased the size of the batch chunk last time). I still think 105 min is reasonable, but I would like to understand the 'lost' time.

**Comment by malaffoon** [comment]
My guess would be the overhead (in the batch framework) of instantiating and tearing down the steps. Imagine that getting a connection takes a couple seconds -- from the framework's point of view, does that happen during a step or while setting up the step?

**Comment by agorina** [comment]
Just to note: I re-ran it with 1 mil as a batch size (double from the before) and the execution time was 62 min.
